modules/imgproc/test/test_thresh.cpp

```diff
@@ -520,6 +520,81 @@ TEST(Imgproc_Threshold, threshold_dryrun)
+
+typedef testing::TestWithParam< Imgproc_Threshold_Masked_Params_t > Imgproc_Threshold_Masked;
+
+TEST_P(Imgproc_Threshold_Masked, threshold_mask)
+{
+    bool useROI = get<0>(GetParam());
+    int depth = get<1>(GetParam());
+    int cn = get<2>(GetParam());
+    int threshType = get<3>(GetParam());
+    int threshFlag = get<4>(GetParam());
+    int type = CV_MAKETYPE(depth, cn);
+    const bool isValidConfig =
+      ( (threshFlag != THRESH_OTSU) && (threshFlag != THRESH_TRIANGLE) ) ||
+      ( (threshFlag == THRESH_OTSU) && ( (type == CV_8UC1) || (type == CV_16UC1) ) ) ||
+      ( (threshFlag == THRESH_TRIANGLE) && (type == CV_8UC1) );
+    if (isValidConfig)
+    {
+        const int _threshType = threshType | threshFlag;
+        Size sz(127, 127);
+        Size wrapperSize = useROI ? Size(sz.width+4, sz.height+4) : sz;
+        Mat wrapper(wrapperSize, CV_MAKETYPE(depth, cn));
+        Mat input = useROI ? Mat(wrapper, Rect(Point(), sz)) : wrapper;
+        cv::randu(input, cv::Scalar::all(0), cv::Scalar::all(255));
+
+        const bool isAutoThreshold = (threshFlag == THRESH_OTSU) || (threshFlag == THRESH_TRIANGLE);
+        if (isAutoThreshold)
+        {
+            //to compare results with OTSU and TRIANGLE, we use a rectangular ROI so that it can be compared
+            //with the non-masked version
+            Mat mask = cv::Mat::zeros(sz, CV_8UC1);
+            cv::Rect roiRect(sz.width/4, sz.height/4, sz.width/2, sz.height/2);
+            cv::rectangle(mask, roiRect, cv::Scalar::all(255), cv::FILLED);
+
+            Mat output_with_mask = cv::Mat::zeros(sz, input.type());
+            const double autoThreshWithMask = cv::thresholdWithMask(input, output_with_mask, mask, 127, 255, _threshType);
+            output_with_mask = Mat(output_with_mask, roiRect);
+
+            Mat output_without_mask;
+            const double autoThresholdWithoutMask = cv::threshold(Mat(input, roiRect), output_without_mask, 127, 255, _threshType);
+
+            ASSERT_EQ(autoThreshWithMask, autoThresholdWithoutMask);
+            //EXPECT_MAT_NEAR(output_with_mask, output_without_mask, 0);
+        }
+        else//if (!isAutoThreshold)
+        {
+            Mat mask = cv::Mat::zeros(sz, CV_8UC1);
+            cv::RotatedRect ellipseRect((cv::Point2f)cv::Point(sz.width/2, sz.height/2), (cv::Size2f)sz, 0);
+            cv::ellipse(mask, ellipseRect, cv::Scalar::all(255), cv::FILLED);//for very different mask alignments
+
+            Mat output_with_mask = cv::Mat::zeros(sz, input.type());
+            cv::thresholdWithMask(input, output_with_mask, mask, 127, 255, _threshType);
+
+            cv::bitwise_not(mask, mask);
+            input.copyTo(output_with_mask, mask);
+
+            Mat output_without_mask;
+            cv::threshold(input, output_without_mask, 127, 255, _threshType);
+            input.copyTo(output_without_mask, mask);
+
+            EXPECT_MAT_NEAR(output_with_mask, output_without_mask, 0);
+        }
+    }
```

Comment:
If-else branches use different pattern and a bit different logic. I propose to split them on 2 test cases with different names. It also allows to get rid of complex logic with `isValidConfig` flag.